src/wp-includes/webfonts-api/class-wp-webfonts-controller.php

```diff
@@ -0,0 +1,232 @@
+		$providers = $this->providers_registry->get_all_registered();
+		if ( empty( $providers ) ) {
+			return '';
+		}
+
+		$styles = '';
+
+		/*
+		 * Loop through each of the providers to get the CSS for their respective webfonts
+		 * to incrementally generate the collective styles for all of them.
+		 */
+		foreach ( $providers as $provider_id => $provider ) {
+			$registered_webfonts = $this->webfonts_registry->get_by_provider( $provider_id );
+
+			// If there are no registered webfonts for this provider, skip it.
+			if ( empty( $registered_webfonts ) ) {
+				continue;
+			}
+
+			/*
+			 * If the provider is external (fetches from an external font service),
+			 * remove any registered webfonts that are not marked as "okay to fetch"
+			 * from the external service.
+			 */
+			if ( $provider->is_external() ) {
+				$registered_webfonts = array_filter(
+					$registered_webfonts,
+					static function( $webfont ) {
+						return ( isset( $webfont['is-external'] ) && true === $webfont['is-external'] );
+					}
+				);
+
+				/*
+				 * After checking if each registered webfont is okay to fetch from external service,
+				 * if there are no remaining webfonts to process, skip this provider.
+				 */
+				if ( empty( $registered_webfonts ) ) {
+					continue;
+				}
+			}
+
+			/*
+			 * Process the webfonts by first passing them to the provider via `set_webfonts()`
+			 * and then getting the CSS from the provider.
+			 */
+			$provider->set_webfonts( $registered_webfonts );
+			$styles .= $provider->get_css();
+		}
+
+		return $styles;
+	}
+
+	/**
+	 * Gets the resource hints.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array  $urls {
+	 *     Array of resources and their attributes, or URLs to print for resource hints.
+	 *
+	 *     @type array|string ...$0 {
+	 *         Array of resource attributes, or a URL string.
+	 *
+	 *         @type string $href        URL to include in resource hints. Required.
+	 *         @type string $as          How the browser should treat the resource
+	 *                                   (`script`, `style`, `image`, `document`, etc).
+	 *         @type string $crossorigin Indicates the CORS policy of the specified resource.
+	 *         @type float  $pr          Expected probability that the resource hint will be used.
+	 *         @type string $type        Type of the resource (`text/html`, `text/css`, etc).
+	 *     }
+	 * }
+	 * @param string $relation_type The relation type the URLs are printed for,
+	 *                              e.g. 'preconnect' or 'prerender'.
+	 * @return array URLs to print for resource hints.
+	 */
+	public function get_resource_hints( $urls, $relation_type ) {
+		foreach ( $this->providers_registry->get_all_registered() as $provider ) {
+			foreach ( $provider->get_resource_hints() as $relation => $relation_hints ) {
+				if ( $relation !== $relation_type ) {
+					continue;
+				}
+				foreach ( $relation_hints as $hint ) {
+					$urls[] = $hint;
+				}
```

Comment:
```suggestion
				$urls = array_merge( $urls, $relation_hints );
```

Reply:
`array_merge()` will create a new array. If the original array is large, this could add processing time and memory. The goal is not to create a new array, but rather add the new items onto the end of the original `$urls` array. `array_push()`  with splat operator would work:

```suggestion
				array_push( $urls, ...$relation_hints );
```

See it in action here https://3v4l.org/oFOlS.

Reply:
Added in this commit https://github.com/WordPress/wordpress-develop/pull/1736/commits/a16c72bef59566047a41dc132e5ce09a4207ceab.